client/task_runner_test.go

```diff
@@ -195,3 +198,65 @@ func TestTaskRunner_SaveRestoreState(t *testing.T) {
 	}
 }
 */
+
+func TestTaskRunner_HandleDiscovery(t *testing.T) {
+	// Create the task runner
+	_, tr := testTaskRunner()
+
+	// Modify the job so we have predictable results
+	tr.allocID = "1"
+	tr.jobID = "job1"
+	tr.taskGroup = "group1"
+	tr.task.Name = "web"
+
+	// Assign dynamic ports. The ReservedPorts are overloaded for dynamic
+	// ports also, so we set those here as well.
+	tr.task.Resources.Networks[0].DynamicPorts = []string{"http", "https"}
+	tr.task.Resources.Networks[0].ReservedPorts = []int{80, 443}
+
+	// Create the mock discovery layer
+	providers := []discovery.Factory{discovery.NewMockDiscovery}
+	disc, err := discovery.NewDiscoveryLayer(providers, tr.config, tr.logger, tr.config.Node)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	tr.discovery = disc
+	mp := disc.Providers[0].(*discovery.MockDiscovery)
+
+	// Register the task with discovery
+	tr.handleDiscovery(false)
+
+	// Check the result
+	expect := map[string]int{
+		"job1.group1.web:1":       0,
+		"job1.group1.web.http:1":  80,
+		"job1.group1.web.https:1": 443,
```

Comment:
> I'd love to manually input at my jobfile what IP(optional), and ports (optional) I want to register at consul for my task. 

@zenvdeluca Yes, as you noted, my intent with the `service{}` block was to be able to specify some additional options. For example, you may not want to advertise all of the ports your service uses, even though you need to bind them. We could do something like.
```hcl
service "http" {
  port = "8080"
  advertise = false # Don't register with service discovery
}
```
I think we can support this for IP via subnet filtering. E.g. only bind / advertise to an interface matching `10.0.0.0/8`, but since this should only happen at the allocation level, you would not need to specify it again in the service block.

As I recall the design specifies the entire task group is allocated on a single interface. I have made an issue to follow up with this here: https://github.com/hashicorp/nomad/issues/264